### ISSUE TITLE
Get rid of 'instanceof' in v2 (or in general?)

### ISSUE THREAD

**Issue by FSMaxB** [josdejong/mathjs#345]
I've seen many lines of code like this:
```js
if (node instanceof SymbolNode) {
  ... ...
}
```

This can be problematic because it isn't guaranteed that the node has been created from the same SymbolNode.

The same goes for types and other classes/prototypes or whatever they are called.

I suggest introducing a `type` property as the nodes already have and test against that instead.

The only place where instanceof should be allowed is in constructors to test if the `new` operator has been used.

**Comment by josdejong** [comment]
I totally agree with the problems using `instanceof`. As for the Nodes, they already have a `type` property so that could be easy.

I was under the impression that checking a property like `if (x && x.type === 'MyType')` was way slower than `if (x instanceof MyType)`, but that may not be the case. A naive test of mine in `v2` resulted in the `add` function being four times slower, but that may have an other cause. In the `develop` branch the function kept the same performance. Need to do some more benchmarking there.

**Comment by FSMaxB** [comment]
Instead of comparing  strings it would certainly be faster  to  define a boolean variable (I'm doing this to replace the  `isNode` function at the moment)

```js
Node.isNode = true
```

The types could  then have their own  booleans,  like  `BigNumber.isBigNumber = true` and so on.

**Comment by josdejong** [comment]
Nice, great idea!! And for example for matrices you could have two properties: `isMatrix` from the base class, and `isDenseMatrix` or `isSparseMatrix` for the implementations. I still have to verify whether there are no performance drawbacks, but I don't expect so (I think my earlier performance issue was caused by an optimization problem in `typed-function`, not by the type check itself). Would be great to get rid of all `instanceof` checks.

**Comment by FSMaxB** [comment]
Btw. how does `typed-function` do it's type checks? Doesn't it use `instanceof`, too?

**Comment by josdejong** [comment]
For typed-function, you have to provided type checks for your own types yourself. Right now you have a check like:

```js
typed.types['Complex'] = function (x) { 
  return x instanceof math.type.Complex; 
};
```

And we can easily replace this with:
```js
typed.types['Complex'] = function (x) { 
  return x && x.isComplex; 
};
```

See code: https://github.com/josdejong/mathjs/blob/v2/lib/util/typed.js#L14-L21

**Comment by josdejong** [comment]
I've now removed all instanceof checks in math.js (except in prototype constructors). Performance issue still needs to be resolved for function `add` and some others, but I'm quite certain that this is related to undefined ordering of types in `typed-function`, which is not very hard to solve.

I have created all type checks like `x && x.isComplex === true` (with the explicit `=== true`). I'm not sure if that's really needed and better or worse for performance.
